### PR TITLE
Add ISoundService.GetInstalledSoundAsync

### PR DESCRIPTION
**Pull request by Sergio0694**
This PR adds `ISoundService.GetInstalledSoundAsync`, as discussed on Discord:

```csharp
namespace AmbientSounds.Services;

public interface ISoundService
{
    Task<Sound?> GetLocalSoundAsync(string soundId);
}
```

### CODE REVIEW

File: src/AmbientSounds/Services/SoundService.cs [hunk 1]
```diff
@@ -39,6 +37,14 @@ public class SoundService : ISoundService
         _assetsReader = assetsReader;
     }
 
+    /// <inheritdoc/>
+    public async Task<Sound?> GetLocalSoundAsync(string soundId)
+    {
+        Guard.IsNotNull(soundId);
```
Comment (dpaulino) [inline src/AmbientSounds/Services/SoundService.cs:43]: Let's return null instead of throwing

**Review by dpaulino — CHANGES_REQUESTED** [review]

File: src/AmbientSounds/Cache/SoundCache.cs [hunk 2]
```diff
@@ -46,6 +46,13 @@ public async Task<IReadOnlyList<Sound>> GetPreinstalledSoundsAsync()
         return _preinstalled.Values as IReadOnlyList<Sound> ?? Array.Empty<Sound>();
     }
 
+    /// <inheritdoc/>
+    public async Task<Sound?> GetInstalledSoundAsync(string stringId)
+    {
+        await GetInstalledSoundsAsync();
+
+        return _installedSounds.TryGetValue(stringId, out Sound result) ? result : null;
+    }
```
Comment (dpaulino) [inline src/AmbientSounds/Cache/SoundCache.cs:55]: why was this moved?

File: src/AmbientSounds/Services/SoundService.cs (hunk 1 above)
Comment (Sergio0694) [reply src/AmbientSounds/Services/SoundService.cs:43]: Why? If you pass a `null` id to a method that specifically needs an id to get a sound that looks like user error to me 🤔

File: src/AmbientSounds/Cache/SoundCache.cs (hunk 2 above)
Comment (Sergio0694) [reply src/AmbientSounds/Cache/SoundCache.cs:55]: This way it's right above the version taking an array of ids, figured it made sense to keep the two close together.

File: src/AmbientSounds/Services/SoundService.cs (hunk 1 above)
Comment (dpaulino) [reply src/AmbientSounds/Services/SoundService.cs:43]: Avoids the need to have a try catch to surround this method when it's called. Null is already a valid return value, so the calling method is already handling the null scenario correctly. So an exception does not seem necessary. What use does it offer in this scenario?

Comment (Sergio0694) [reply src/AmbientSounds/Services/SoundService.cs:43]: > "Avoids the need to have a try catch to surround this method when it's called"

You should definitely not do that, that is not the correct way to handle this kind of errors. Passing `null` to a method with a non nullable reference type is a programming error, not an exceptional error. You should not be adding try/catch blocks for the former, only for the latter. For the former, you should rather fix the code so that those don't happen. In this scenario, you'd just make sure that callers never pass a `null` values in the first place (which the annotations can help you with).

Comment (dpaulino) [reply src/AmbientSounds/Services/SoundService.cs:43]: I understand where you're coming from. The parameter should never be null (because it's declared as string instead of string?) and the guard just enforces that. 

But what advantage does it offer in this scenario over returning null?

Comment (Sergio0694) [reply src/AmbientSounds/Services/SoundService.cs:43]: It makes the contract clear and enforces it, which pushes consumers to use it correctly, as opposed to silently ignoring an invalid condition, which might hide other problems upstream (if someone did pass null here it'd mean something else has gone wrong) and make it harder to investigate and resolve them. I'm strongly convinced that the fail-fast approach is fundamentally superior and a better investment in the long run, whereas just ignoring error conditions only makes the codebase more brittle and less reliable over time. This is just a very small example of this 🙂

TLDR: this avoids those "mmmh this isn't working, no idea why" moments later on.

Comment (dpaulino) [reply src/AmbientSounds/Services/SoundService.cs:43]: I understand you believe the fail fast principle is superior, but in practice, I think the extreme end of this principle leads to unnecessary difficulties for ambie customers and for me as the app owner. 

Let's imagine that crash happened in real life.
- Customers would get a crash in a certain scenario.
- I may get poor ratings because of the crash
- Users might stop using ambie entirely because they perceive it as unstable
- How do I know a crash is happening for that edge case? A customer might have to tell me, or I check my telemetry. 
- How long might it take for me to figure out there's a crash in an edge case? In practice, could be between one to 7 days because this is just a hobby project for me. 
- Then I go fix the issue caused by this fail fast crash, then I submit the fix to alpha to confirm the fix. Then I submit to retail. 
- How long does it take to go to retail? Maybe 24 hours after I fix the issue. 
- How long does it take to be deployed to customers? Well it usually takes about 5 days for the latest app version to become the most actively used version based on my telemetry. This is because windows update does not update store apps instantly after a new release is available. 
- But if a user already stopped using ambie because of a crash, they would never use ambie again with the fix.

To summarize:
- For a windows application like ambie, the cost of a user facing crash is very high due to the update process. 
- It leads to lost customers and loss of customer trust
- It can take at least 24 hours to get a fix out, and that's ignoring the time it takes for me to realize the issue and fix it, and ignores the time it takes for the update to propagate to all devices. 

So the fail fast method in the extreme is not the ideal for ambie. I do not believe that an app like ambie should be ever crash. The UX is more important than design by contract. Returning null in this scenario is fine and does not need to strictly adhere to fail fast. 

If I were building a web service that can be updated on the fly, or a command line tool that usually has a short, transient UX, I can see fail fast being useful. But for an app like ambie to crash, especially a crash that was by design, I don't want my customers to experience that.

Comment (Sergio0694) [reply src/AmbientSounds/Services/SoundService.cs:43]: I think you're missing the point and not seeing the full picture. That crash would _never_ happen in production. The whole point of annotations is to make code correct by construction in cases like this, as you'd get a warning at the callsite if you were trying to pass a value that could be `null`. If you did have a bug, the idea is to simply catch that during development. Those guards are more like asserts, but they should never be hit in practice. Additionally, if a caller was actually trying to pass `null` it'd mean that something went wrong on its side, which would in practice result in stuff like a button not doing anything or similiar, which are still situations that (while less severe than a crash) are not a great UX for consumers anyway. And on top of this you'd have to add the fact that it'd make it significantly harder for you to spot such bugs in your code.

No problems changing this (I mean it's your app), but I would strongly recommend you to take some time to think about it at some point, as you're effectively pushing for a pattern that results in code progressively becoming less reliable🙂

Comment (dpaulino) [reply src/AmbientSounds/Services/SoundService.cs]: question: what if we change the parameter to `string?` and then return null if the id is null? Can we do that instead?

Comment (Sergio0694) [reply src/AmbientSounds/Services/SoundService.cs]: That's what I assumed you were asking for, and I've already changed this in the latest commit 👍
